src/Data/Singletons/Partition.hs

```diff
@@ -46,23 +48,35 @@ partitionDecs :: Quasi m => [DDec] -> m PartitionedDecs
 partitionDecs = concatMapM partitionDec
 
 partitionDec :: Quasi m => DDec -> m PartitionedDecs
+partitionDec (DLetDec (DPragmaD {})) = return mempty
 partitionDec (DLetDec letdec) = return $ mempty { pd_let_decs = [letdec] }
 
 partitionDec (DDataD nd _cxt name tvbs cons derivings) = do
-  (derivings', derived_instances) <- partitionWithM part_derivings derivings
+  (derivings', derived_instances) <- partitionWithM part_derivings
+                                   $ concatMap flatten_clause derivings
   return $ mempty { pd_data_decs = [DataDecl nd name tvbs cons derivings']
                   , pd_instance_decs = derived_instances }
   where
     ty = foldType (DConT name) (map tvbToType tvbs)
-    part_derivings :: Quasi m => DPred -> m (Either DPred UInstDecl)
-    part_derivings deriv = case deriv of
+
+    flatten_clause :: DDerivClause -> [(Maybe DerivStrategy, DPred)]
+    flatten_clause (DDerivClause strat preds) = map (strat,) preds
+
+    part_derivings :: Quasi m => (Maybe DerivStrategy, DPred)
+                              -> m (Either DPred UInstDecl)
+    part_derivings (strat, deriv) = case deriv of
       DConPr deriv_name
-         | deriv_name == ordName
+         | stock, deriv_name == ordName
         -> Right <$> mkOrdInstance ty cons
-         | deriv_name == boundedName
+         | stock, deriv_name == boundedName
         -> Right <$> mkBoundedInstance ty cons
-         | deriv_name == enumName
+         | stock, deriv_name == enumName
         -> Right <$> mkEnumInstance ty cons
+        where
+          stock = case strat of
+                    Nothing            -> True
+                    Just StockStrategy -> True
+                    Just _             -> False
```

Comment:
I opted to only choose `deriving` clauses with an explicit strategy when that strategy is `stock`. We could make it more permissive and ignore the strategy, however.